### PR TITLE
Plugin support

### PR DESCRIPTION
**Pull request by khiron**
Adds plugin-support features to develop branch
- implementation of stevedore library to replace `_app_registry`
- migration of internal apps from automatic registry with `_app_registry` to registration though entry-points registered to the group `cogent3.app`
- conversion of app tests to use the plugin infrastructure 
  - deprecation of tests that dynamically load apps
  - separation of tests that probe composability (into `test_composable.py`) and tests that probe plugin functionality (into `test_plugins.py`)

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -10,4 +10,5 @@ scipy==1.13.0
 nox==2024.4.15
 plotly==5.20.0
 typing_extensions>=3.7
+stevedore
```
Comment (GavinHuttley) [inline requirements.txt:13]: pin the version in here, it lets dependabot trigger PRs when stevedore is updated

File: tests/test_app/test_plugins.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+    with pytest.warns(UserWarning, match=r".* is not a valid cogent3 app, skipping"):
+        mock_extension_manager([create_extension(not_an_app)])
+        apps = available_apps()
+        assert isinstance(apps, Table)
+        apps = apps.filtered(lambda x: not_an_app.__name__ == x, columns="name")
+        assert apps.shape[0] == 1
+
+
+def test_apps_can_be_deprecated(mock_extension_manager):
```
Comment (GavinHuttley) [inline tests/test_app/test_plugins.py:183]: we're deleting this attribute, so remove the tests

File: tests/test_app/test_plugins.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            self.a = a
+
+        def main(self, data: str = "foo") -> str:
+            return data.upper()
+
+    mock_extension_manager([create_extension(DummyApp, module_name="module1")])
+
+    assert DummyApp.__name__ in cogent3.app.get_app_manager().names()
+    dummy_instance = DummyApp()
+    assert (
```
Comment (GavinHuttley) [inline tests/test_app/test_plugins.py:245]: tests of exact string values are fragile, look for certain words. Edit the test to do this.

File: src/cogent3/app/__init__.py [hunk 4]
```diff
@@ -52,36 +56,48 @@ def _make_types(app) -> dict:
     return _types
 
 
+# private global to hold an ExtensionManager instance
+__apps = None
+# exclude apps from deprecated modules
```
Comment (GavinHuttley) [inline src/cogent3/app/__init__.py:61]: delete all uses of the `__deprecated` variable. We don't need this anymore.

**Review by GavinHuttley — CHANGES_REQUESTED** [review]
few small changes, plus whatever we raised in the CZI meeting.

fantastic progress!

**Review by khiron — COMMENTED** [review]
following on from my review during the 23/4 CZI meeting.  All items have been  resolved

**Review by GavinHuttley — APPROVED** [review]
🎉 awesome, ready to merge. You get the privilege of pressing that button 🥳

**Pull request merged**